### PR TITLE
GEH created, @ExceptionHandler used, Logger created, validations added

### PR DESCRIPTION
**Pull request by validatorr**

### CODE REVIEW

File: src/main/java/v/hryhoryk/onlinebookstore/dto/CreateBookRequestDto.java [hunk 1]
```diff
@@ -1,14 +1,27 @@
...
+        @Pattern(regexp = "^(?:ISBN(?:-1[03])?:? )?(?=[0-9X]{10}$|(?=(?:[0-9]+[- ]){3})"
+                + "[- 0-9X]{13}$)[0-9]{1,5}[- ]?[0-9]+[- ]?[0-9]+[- ]?[0-9X]$",
+                message = "Invalid ISBN format")
         String isbn,
+        @Positive(message = "Price can't be negative")
         BigDecimal price,
+        @NotNull(message = "Description should contain some useful information")
         String description,
+        @URL(message =
+                "It's required to add image of book to increase customer's trust rate")
```
Comment (Cayman4010) [inline src/main/java/v/hryhoryk/onlinebookstore/dto/CreateBookRequestDto.java:23]: Hm, it's intersting! I'll know about it.

**Review by Cayman4010 — APPROVED** [review]
Great!

File: pom.xml [hunk 2]
```diff
@@ -64,8 +64,33 @@
...
+            <artifactId>log4j-slf4j-impl</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+        </dependency>
```
Comment (liliia-ponomarenko) [inline pom.xml]: read about logging in SpringBoot https://www.baeldung.com/spring-boot-logging

File: pom.xml [hunk 3]
```diff
@@ -64,8 +64,33 @@
             <artifactId>liquibase-core</artifactId>
             <version>${liquibase.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
+            <artifactId>hibernate-validator</artifactId>
+        </dependency>
```
Comment (liliia-ponomarenko) [inline pom.xml]: it is better to use `spring-boot-starter-validation`

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/BookNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package v.hryhoryk.onlinebookstore.exceptions;
+
+public class BookNotFoundException extends RuntimeException {
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/exceptions/BookNotFoundException.java]: Don't specify your exception. Named it like EntityNotFoundException and you can use it for each entity

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@RestControllerAdvice
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    private static final Logger logger = LogManager.getLogger(CustomGlobalExceptionHandler.class);
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java]: if you want to use logger read article  above 👆

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java]: Does this information have value? The response status will be contained in the response itself, no need to add it to the body

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    private String getErrorMessage(ObjectError e) {
+        if (e instanceof FieldError) {
+            String field = ((FieldError) e).getField();
+            String message = e.getDefaultMessage();
+            return field + " " + message;
+        }
+        return e.getDefaultMessage();
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java]: move private method after public

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    @ExceptionHandler(Exception.class)
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public ErrorResponse handleGlobalException(Exception e) {
+        logger.error("Internal server error: " + e.getMessage(), e);
+        return new ErrorResponse(
+                HttpStatus.INTERNAL_SERVER_ERROR.value(),
+                "Internal server error",
+                e.getMessage()
+        );
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java]: do you rea need handle this exception?

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        if (e instanceof FieldError) {
+            String field = ((FieldError) e).getField();
+            String message = e.getDefaultMessage();
+            return field + " " + message;
+        }
+        return e.getDefaultMessage();
+    }
+
+    @ExceptionHandler(BookNotFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java]: Do you really need duplicate response status?

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    public ErrorResponse handleBookNotFoundException(BookNotFoundException e) {
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java]: Better return  ResponseEntity<Object>

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                HttpStatus.NOT_FOUND.value(),
+                "Book not found",
+                e.getMessage()
+        );
+    }
+
+    @ExceptionHandler(KeySpecificationProviderException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ErrorResponse handleKeySpecificationProviderException(
+            KeySpecificationProviderException e) {
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java]: the same question

File: src/main/java/v/hryhoryk/onlinebookstore/dto/CreateBookRequestDto.java [hunk 12]
```diff
@@ -1,14 +1,27 @@
 package v.hryhoryk.onlinebookstore.dto;
 
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Pattern;
+import jakarta.validation.constraints.Positive;
 import java.math.BigDecimal;
+import org.hibernate.validator.constraints.URL;
 
 @JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/dto/CreateBookRequestDto.java]: Why do you need this annotation?

File: src/main/java/v/hryhoryk/onlinebookstore/dto/CreateBookRequestDto.java [hunk 13]
```diff
@@ -1,14 +1,27 @@
...
 public record CreateBookRequestDto(
+        @NotNull(message = "Title can't be null")
         String title,
+        @NotNull(message = "Author can't be null")
         String author,
+        @Pattern(regexp = "^(?:ISBN(?:-1[03])?:? )?(?=[0-9X]{10}$|(?=(?:[0-9]+[- ]){3})"
+                + "[- 0-9X]{13}$)[0-9]{1,5}[- ]?[0-9]+[- ]?[0-9]+[- ]?[0-9X]$",
+                message = "Invalid ISBN format")
         String isbn,
+        @Positive(message = "Price can't be negative")
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/dto/CreateBookRequestDto.java]: NotNulL ZeroOrPositive

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few mistakes;)

File: src/main/java/v/hryhoryk/onlinebookstore/dto/CreateBookRequestDto.java (hunk 12 above)
Comment (validatorr) [reply src/main/java/v/hryhoryk/onlinebookstore/dto/CreateBookRequestDto.java]: I actually thought that we are not specifying id in our DTO and there will be some kind of conflict while parsing this file by JSON, but now I've read documentation properly and understood that it's unnecessary here

File: src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java (hunk 5 above)
Comment (Andriy-dz) [reply src/main/java/v/hryhoryk/onlinebookstore/exceptions/CustomGlobalExceptionHandler.java]: not fixed

**Review by Andriy-dz — APPROVED** [review]

**Pull request merged**